From 182d586fee5b5f9053b5702fcfc847f18ea5ec79 Mon Sep 17 00:00:00 2001
From: Nick Diego Yamane <nickdiego@igalia.com>
Date: Thu, 27 Feb 2020 02:13:10 +0000
Subject: [PATCH] ozone/x11: Fix an IME-related crash during browser
 initialization

Recent changes in LinuxInputMethodContextFactory initialization (see
crrev.com/c/2062974) broke ozone/x11. After it, browser crashes at
startup. This fixes it as well as adds DCHECK to make sure
LinuxInputMethodContextFactory is correctly set just after LinuxUI
initialization.

Bug: 1055061
Change-Id: Iefe5acc48ec9047b90042b398159e9b260a753f6
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2069229
Commit-Queue: Nick Yamane <nickdiego@igalia.com>
Reviewed-by: Scott Violet <sky@chromium.org>
Reviewed-by: Maksim Sisov <msisov@igalia.com>
Reviewed-by: Robert Kroeger <rjkroege@chromium.org>
Cr-Commit-Position: refs/heads/master@{#744902}
---
 .../views/chrome_browser_main_extra_parts_views_linux.cc   | 3 +++
 ui/ozone/platform/x11/ozone_platform_x11.cc                | 7 +++++++
 2 files changed, 10 insertions(+)

diff --git a/chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc b/chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc
index b6d99cc37f30d..c959e4f66562d 100644
--- a/chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc
+++ b/chrome/browser/ui/views/chrome_browser_main_extra_parts_views_linux.cc
@@ -36,4 +36,7 @@ void ChromeBrowserMainExtraPartsViewsLinux::ToolkitInitialized() {
 
   views::LinuxUI::SetInstance(linux_ui);
   linux_ui->Initialize();
+
+  DCHECK(ui::LinuxInputMethodContextFactory::instance())
+      << "LinuxUI must set LinuxInputMethodContextFactory instance.";
 }
diff --git a/ui/ozone/platform/x11/ozone_platform_x11.cc b/ui/ozone/platform/x11/ozone_platform_x11.cc
index c3e0ec1855aea..07206a3af8efa 100644
--- a/ui/ozone/platform/x11/ozone_platform_x11.cc
+++ b/ui/ozone/platform/x11/ozone_platform_x11.cc
@@ -9,6 +9,7 @@
 
 #include "base/message_loop/message_pump_type.h"
 #include "base/strings/utf_string_conversions.h"
+#include "ui/base/ime/linux/linux_input_method_context_factory.h"
 #include "ui/base/x/x11_util.h"
 #include "ui/display/fake/fake_display_delegate.h"
 #include "ui/events/devices/x11/touch_factory_x11.h"
@@ -114,6 +115,12 @@ class OzonePlatformX11 : public OzonePlatform {
 #if defined(OS_CHROMEOS)
     return std::make_unique<InputMethodChromeOS>(delegate);
 #else
+    // This method is used by upper layer components (e.g: GtkUi) to determine
+    // if the LinuxInputMethodContextFactory instance is provided by the Ozone
+    // platform implementation, so we must consider the case that it is still
+    // not set at this point.
+    if (!ui::LinuxInputMethodContextFactory::instance())
+      return nullptr;
     return std::make_unique<InputMethodAuraLinux>(delegate);
 #endif
   }
